src/seedu/addressbook/data/person/Contact.java

```diff
@@ -0,0 +1,50 @@
+package seedu.addressbook.data.person;
+
+import seedu.addressbook.data.exception.IllegalValueException;
+
+/**
+ * Represents a Person's generic contact information in the address book.
+ * Guarantees: immutable;
+ */
+public abstract class Contact {
+
+    public final String value;
+    private boolean isPrivate;
+
+    public Contact(String contact, boolean isPrivate) {
+        this.isPrivate = isPrivate;
+        this.value = contact.trim();
+    }
+
+    /**
+     * Validates value against regex.
+     *
+     * @throws IllegalValueException if value is invalid.
+     */
+    public void validate(String validationRegex, String messageConstraint) throws IllegalValueException {
+        if (!value.matches(validationRegex)) {
+            throw new IllegalValueException(messageConstraint);
+        }
+    }
+
+    @Override
+    public String toString() {
+        return value;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other.getClass().isInstance(this) // instanceof handles nulls
+                && this.value.equals(((Contact)other).value)); // state check
```

Comment:
Note that now an Email and Address having the same value will be considered equal even if they are of different types.